### ISSUE TITLE
DESeq2 Wrapper is a bit opaque in how it works

### ISSUE THREAD

**Issue by lparsons** [galaxyproject/tools-iuc#716]
I'd like to understand how the Galaxy wrapper for DESeq2 translates the factors and factor levels into an experimental design and how the contrasts are setup. A quick look for the R code turned up nothing since the script used is embedded somewhere in a dependency. I have a few suggestions:
1. Add information to the `help` section of the tool that explains how the experimental design is created from various factors (it's particularly unclear how multiple factors are handled).
2. Add information to the `help` section of the tool that explains how the contrasts are generated (which factors and levels, etc.). Again, this is especially important in multifactorial designs.
3. Add headers to the output tables that describe the design formula, the contrast, and include a header for each column.

**Comment by nekrut** [comment]
Yes. This needs clarification, but take a look at this example by @bgruening : https://github.com/bgruening/training-material/blob/master/rna-seq/rna-seq.md

**Comment by lparsons** [comment]
Thanks @nekrut, I'll take a look. Another issue is that most of this tool is actually in scripts that are hosted in the DESeq2 package on galaxydepot. Is that code available in a public repo somewhere? Any work on that tool would require updating that code...

**Comment by nekrut** [comment]
@bgruening ?

**Comment by mvdbeek** [comment]
> Another issue is that most of this tool is actually in scripts that are hosted in the DESeq2 package on galaxydepot. Is that code available in a public repo somewhere?

It's an ongoing discussion, it used to be in the repo, now it lives in the DESeq2 source:
https://github.com/Bioconductor-mirror/DESeq2/blob/master/inst/script/deseq2.R

About the headers, I believe the goal is to have them reported as metadata in galaxy starting release_16.04.

**Comment by lparsons** [comment]
@mvdbeek Thanks, good to know the scripts are somewhere. Also, very excited for annotated output using metadata. Can't tell you how many people ask me what the columns of things are, and it's just not intuitive to look at the tool help, since that is where you go to run the thing and it's really connected to the output. Though that gives me an idea... link to tool help from tool output in history?

I know this is a larger issues, but it would be great to expose those things hosted on `galaxydepot` somewhere and show how they are built. I worry that Galaxy installations are increasingly dependent on ongoing support from the Galaxy team due to things like this.

**Comment by mvdbeek** [comment]
> I know this is a larger issues, but it would be great to expose those things hosted on galaxydepot somewhere and show how they are built

galaxydepot (for the most part) is just a mirror, in case the original source download goes away. A few big packages (R for example) are also hosted as binary versions, so admins don't spend hours compiling these. Everyone can get things into galaxydepot, the procedure is outlined here:
https://github.com/galaxyproject/cargo-port under contributions.

I do agree with your general notion, nowadays you need to know a lot of different things and details to understand exactly where a dependency is coming from.

**Comment by bgruening** [comment]
Hi all,

sorry to be so late to this discussion, I'm on a workshop these days. 
A few points:

* I'm fine with moving the DESeq2 script back into Galaxy's wrapper. It used to be there and we can move it back if @mikelove does not need it in his repo, which seems to be the case.
* the script is currently available one the BioC github mirror and available from BioC, so it is not hidden at all
* the final wrapper will have support for metadata column names and will follow the 16.04 tool specifications
* I disagree that dependencies are getting more complicated. Essentially `depot` is a mirror for tarballs. If we store binaries or wheels there, all the recipes are generated with Starforge and on github, so completely transparent. Conda will hopefully replace tool_dependencies and Starforge soon, so in the end you have only Conda packages. The source from Conda packages are mirrored to depot and we can build automatic Docker containers from Conda. So the flow is like this:

depot <- tarball -> Conda -> Docker 

@lparsons feel free to enhance the help section of DESeq2, I always feel that it's hard to explain this in a correct but easy to understand way.

**Comment by lparsons** [comment]
@bgruening First off, I want to say that I completely understand the value of the `depot` in terms of practicality and making things "just work". For that reason, I think it's likely worth the tradeoffs. However, I do think there are tradeoffs in complexity (it's at least a mirror and point of failure), and in transparency.

I have a few issues that I'd like to see addressed (and I'd be willing to help where I can).

1. When I look at a tool in the toolshed, I see that it's downloading some tarball from `depot`. I can't (easily) determine what that code is, where it came from, how it was generated, etc. In this instance, I had no way of knowing it was an unmodified tarball from bioconductor. (Is that even correct?)

2. In addition to not really know what the code is, I have no simple way to contribute updates or fixes to it. This is a piece that is potentially separate from either the tool in the toolshed or the upstream tool itself. BTW, in this case, I think I'd have a preference to keep the Galaxy specific code in the Galaxy toolshed repository and not a separate (third) repo.

3. My Galaxy installation is now dependent on the `depot` being up and maintained by the Galaxy team. While this is a nice service, it's an added level of dependency that wasn't there when the recipes simply got the source code directly from the upstream project.

I do believe that documentation could go a long way to mitigating these issues. Perhaps adding information to the wiki as well as a landing page on https://depot.galaxyproject.org/ that helps people to understand what is there and how it was generated.

In short, I very much appreciate everyone's efforts and I do think things are getting more reliable, but from my perspective they are getting increasingly complex and the documentation of how things work, how things are changing, and where things are going is lacking. For example, conda, docker, depot, and starforge are all things I've heard of, but don't fully grasp how the Galaxy Project is using them.

**Comment by lparsons** [comment]
@bgruening Oh, and as far as the DESeq2 help section, I'm happy to help, but first I need to sort out what it's actually doing. I still am not quite sure (bit of code to dig through) ;-)

**Comment by yhoogstrate** [comment]
@lparsons It's a bit of a hidden URL but I think you're looking for these guys:
https://depot.galaxyproject.org/software/
https://github.com/galaxyproject/cargo-port/blob/master/urls.tsv

**Comment by lparsons** [comment]
@yhoogstrate Thanks! Those are extremely helpful. What we need is an overview document that describes how everything fits together. Probably needs to be linked directly from the main toolshed.

**Comment by bgruening** [comment]
> @bgruening First off, I want to say that I completely understand the value of the `depot` in terms of practicality and making things "just work". For that reason, I think it's likely worth the tradeoffs. However, I do think there are tradeoffs in complexity (it's at least a mirror and point of failure), and in transparency.

Shouldn't be if documented well, here we failed obviously.

> 1. When I look at a tool in the toolshed, I see that it's downloading some tarball from `depot`. I can't (easily) determine what that code is, where it came from, how it was generated, etc. In this instance, I had no way of knowing it was an unmodified tarball from bioconductor. (Is that even correct?)

Yes, this is correct. You find this information in the depot github site. It's all contained in a tsv file.
We wanted to make this more transparent and here is a little bit of history and an outline what we wanted to do:

https://gist.github.com/bgruening/48297c27cd72cbadea7a

> 2. In addition to not really know what the code is, I have no simple way to contribute updates or fixes to it. This is a piece that is potentially separate from either the tool in the toolshed or the upstream tool itself. BTW, in this case, I think I'd have a preference to keep the Galaxy specific code in the Galaxy toolshed repository and not a separate (third) repo.

Mh this one I does not understand. You can just add a newline to the cargo-port tsv file and can store a new tarball. Changing tarballs is not the purpose of cargo-port. This is a really stupid store. It just mirrors source. It really only solves the problem that many tarballs just disappear over time.

> 3. My Galaxy installation is now dependent on the `depot` being up and maintained by the Galaxy team. While this is a nice service, it's an added level of dependency that wasn't there when the recipes simply got the source code directly from the upstream project.

In the gist we explained how we want to first look at the original source of the source and only in case of a 404 look at depot. Moreover, I trust depot more than some random other FTP server, but this might be just me :) 

> I do believe that documentation could go a long way to mitigating these issues. Perhaps adding information to the wiki as well as a landing page on https://depot.galaxyproject.org/ that helps people to understand what is there and how it was generated.

I'm willing to write something about this and maybe you can improve it until it is understandable :)
I'm just not sure if this worth the effort, actually all this will not be needed in a few month if everything goes well and we are using conda (and Docker) packages for everything.

> In short, I very much appreciate everyone's efforts and I do think things are getting more reliable, but from my perspective they are getting increasingly complex and the documentation of how things work, how things are changing, and where things are going is lacking. For example, conda, docker, depot, and starforge are all things I've heard of, but don't fully grasp how the Galaxy Project is using them.

Sorry for that. I'm aware of the increased complexity of all these new technologies, it's a rough time as we are in a transition phase. I'm happy to document all this or give an outline of how imagine things will work in a few month. If everything goes well all what we need to care about is a conda package and the rest we will get for free as `depot` and docker will be integrated upstream - into the various conda repositories.

@lparsons happy to chat with you about that, maybe you can help me to write this down.

**Comment by mvdbeek** [comment]
The script now lives with the deseq2 tool. There was some talk of eventually moving the script and related scripts into a package, but until a volunteer steps up that won't happen.
